lib/idle_task_queue.dart

```diff
@@ -0,0 +1,88 @@
+// @JS()
+library idle_task_queue;
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:js';
+
+import 'package:meta/meta.dart';
+
+const Duration _immediately = const Duration();
+
+/// A handler which will be called during idle time.
+typedef void IdleTaskHandler<T>(T input);
+
+@immutable
+class _IdleTask<T> {
+  final IdleTaskHandler<T> handler;
+  final T input;
+  final Completer<Null> completer = new Completer<Null>();
+
+  _IdleTask(this.handler, this.input);
+}
+
+/// Adds the task to a queue which will be called during idle time.
+///
+/// Returns the `Future` which will complete when the task has run.
+///
+/// May complete with an error if the task throws when called.
+Future<Null> enqueueIdleTask<T>(IdleTaskHandler<T> handler, T input) {
```

Comment:
What are your thoughts on passing through the result of the call to `handler`?

```dart
typedef FutureOr<R> IdleTaskHandler<T, R>(T input);
Future<R> enqueueIdleTask(IdleTaskHandler<T, R> hander, T input) { ... }
```

I'm thinking that might make it easier to convert existing async logic to use this utility.
```dart
Future<String> doTheThingAsync() async => 'the thing';
String doTheThing() => 'the thing';

final theThing1 = await enqueuIdleTask((_) => doTheThingAsync(), null);
final theThing2 = await enqueuIdleTask((_) => doTheThing(), null);
```